### ISSUE TITLE
[Python] Add local vol model

### ISSUE THREAD

**Issue by ccjeremylo** [KYLChiu/the-big-learning-repo#52]
See #14 for background.

# Proposal for local vol model design:

### Assume we have access to liquid put and call option quotes

- Propose using a [panda dataframe](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) as container for the discrete set of put and call price quotes for a range of $T$ and $K$.
- For the purpose of this "library", we will use some dummy data for testing/analysis, unless someone knows any cheap and easy way to access option quotes
- In practise, it is not clear to me if market option prices are quoted in terms of [IV](https://en.wikipedia.org/wiki/Implied_volatility) or [option premium](https://www.investopedia.com/articles/optioninvestor/09/buying-options.asp#:~:text=Options%20prices%2C%20known%20as%20premiums,premium%20above%20its%20intrinsic%20value.). Our infrastructure should be able to handle both. Conversion from option premium to IV is available thanks to #16.

### Code up a `vol_surface_interpolation` abstract base class:

- has a virtual function `interpolate()`
- write a naive linear 2D interpolation child class for now - very [bad idea](https://quant.stackexchange.com/questions/20741/why-linear-interpolation-not-appropriate-for-volatility-surface-construction)!
- add [spline based](https://en.wikipedia.org/wiki/Spline_interpolation) methods later

### Code up an `implied_vol_surface` abstract base class

- has a virtual function `implied_vol()` to get IV any given T and K: $T,K \mapsto \sigma(T-t_0,K)$
- write an `equity_BS_implied_vol_surface` child class that takes in object `vol_surface_interpolation` as input in ctor

### Code up a `local_vol_surface` abstract base class

- has a virtual function `local_vol()`
- write  `BS_local_vol` child class that takes in object `equity_BS_implied_vol_surface` as input in ctor
- see eq 2.19 of [Bergomi](https://www.lorenzobergomi.com/contents-sample-chapters) - equation can be found in ch.2
- any [LSV](https://link.springer.com/chapter/10.1057/9781137335722_9) model we add in the future would be a child class of `local_vol_surface`

### Use local vol given by `BS_local_vol` object in our GBM analytical stock price long stepping simulator

- by long stepping GBM, I mean solving GBM SDE analytically, ie $\ln \frac{S(T)}{S(t_0)}= \int_{t_0}^T r(s) - q(s) - \frac{1}{2} \widetilde{\sigma}^2(T-s,K) ds + \int_{t_0}^T \widetilde{\sigma}(T-s,K) dW_s $, where $T$ is option maturity and $\widetilde{\sigma}(\tau,K)$ is the local vol for time to expiry $\tau$ and option strike $K$ - I think here we are implicitly assuming [sticky strike](https://en.wikipedia.org/wiki/Volatility_smile#Evolution:_Sticky) rolling of the LV surface...more thinking required
- In practise, given our current set up (of using the [parameter class](https://github.com/KYLChiu/ExoticMonteCarloEngine/blob/master/python/ExoticEngine/MarketDataObject/Parameter.py)) we would likely need to write a `lambda` $s \mapsto \widetilde{\sigma}(s,K)$ for a given $K$, and use that as an input to our [BS simulation engine](https://github.com/KYLChiu/ExoticMonteCarloEngine/blob/master/python/ExoticEngine/MonteCarloEngine/SimulationModel.py) - doesn't seem like a scalable design...suggestions are welcome
- Need some thinking how to handle discrete steps, e.g. [Euler method](https://en.wikipedia.org/wiki/Euler%E2%80%93Maruyama_method)

### Write tests and notebook to check:

- Smile dynamics - needs some thinking, see ch.2.5 of [Bergomi](https://www.lorenzobergomi.com/contents-sample-chapters)
- Ability to reprice OTM and ITM vanilla options
- This [LSE deck](http://www.maths.lse.ac.uk/Personal/jruf/teaching/1206_StochVolSlides.pdf) gives quite a good overview for smile/vol modelling without going into crazy maths

**Comment by KYLChiu** [comment]
Generally LGTM - minor comments/questions only:

Re "Assume we have access to liquid put and call option quotes": not sure if we mean store quotes in a csv file or equivalent. Could run out of memory if we have an especially large dataset. Dummy data is good, we probably want to minimise dependencies on other libraries anyway (as much as possible).

For vol_surface_interpolation: does interpolate need to take 4 parameters? (t0, K0, t1, K1)? Might be better to group these into tuples if so?

Agree that parameter class lambdas should be limited... anything against EM time-stepping?

**Comment by ccjeremylo** [comment]
> 

Yes, exactly, I meant storing quotes in a csv format. Definitely not some large dataset. Not supricingly, lots of people  have worked on vol surface modelling, and I managed to find some open source analysis on vol surface modelling with real option data available - we can borrow the data from there when needed - see e.g. [here](https://github.com/cantaro86/Financial-Models-Numerical-Methods/blob/master/data/spy-options-exp-2020-07-10-weekly-show-all-stacked-07-05-2020.csv). His github profile has a lot of interesting work in it actually.

For the `vol_surface_interpolation`, I am assuming we have access to the entire vol surface. I think I can send out a PR shortly actually - would very much welcome any code design feedback.

No, nothing about EM-stepping, just that I have not put too much thoughts into it. Start with something easy first, given we know the exact solution of GBM. Why we start doing fancier models like SV or jumps etc, then we can't get away with using numerical methods. The interface should be identical anyway.
